File: src/main/java/com/arclights/musiclights/core/filter/ByGroup.java
package com.arclights.musiclights.core.filter;

import ddf.minim.analysis.FFT;

import java.util.ArrayList;

public class ByGroup {

    public static ArrayList<Float> filter(FFT fft, int nbrOfGroups) {
        ArrayList<Float> res = new ArrayList<Float>();
        int spectrumWidth = fft.specSize() / 2;
        int bandsPerGroup = (int) (spectrumWidth / nbrOfGroups);
        for (int i = 0; i < nbrOfGroups; i++) {
            float largestAmpInGroup = 0;
            for (int j = i * bandsPerGroup; j < (i + 1) * bandsPerGroup; j++) {
                if (fft.getBand(j) > largestAmpInGroup) {
                    largestAmpInGroup = fft.getBand(j);
                }
            }
            res.add(largestAmpInGroup);
        }

        return res;
    }

}

File: src/main/kotlin/com/arclights/musiclights/core/filter/ByGroup.kt
package com.arclights.musiclights.core.filter

import ddf.minim.AudioInput
import ddf.minim.analysis.FFT

class ByGroup(private val nbrOfGroups: Int) : Filter {
    override fun filter(fft: FFT, input: AudioInput): List<Float> {
        val spectrumWidth = fft.specSize() / 2
        val bandsPerGroup = spectrumWidth / nbrOfGroups
        return (0 until nbrOfGroups)
            .map {
                (it * bandsPerGroup until (it + 1) * bandsPerGroup)
                    .map(fft::getBand)
                    .max()!!
            }
    }
}